File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def solution(s):
    answer = 0
    queue = deque(s)
    
    x = len(s) # 회전 수
    
    stack = []
    
    for q in queue:
        if len(stack) == 0: # 스택이 비어있을 때
            stack.append(q)
            continue
        
        # 스택의 마지막 원소와 큐의 원소 비교
        if stack[-1] == '(' and q == ')':
            stack.pop()
        elif stack[-1] == '[' and q == ']':
            stack.pop()
        elif stack[-1] == '{' and q == '}':
            stack.pop()
        else:
            stack.append(q)
    
    if len(stack) == 0:
        answer += 1
    
    for i in range(1, x):
        stack = []
        queue.rotate(-1)
        for q in queue:
            if len(stack) == 0: # 스택이 비어있을 때
                stack.append(q)
                continue

            # 스택의 마지막 원소와 큐의 원소 비교
            if stack[-1] == '(' and q == ')':
                stack.pop()
            elif stack[-1] == '[' and q == ']':
                stack.pop()
            elif stack[-1] == '{' and q == '}':
                stack.pop()
            else:
                stack.append(q)

        if len(stack) == 0:
            answer += 1
    
    return answer

"""
## 괄호 회전하기

## 문제
1. (), [], {} => 올바른 괄호 문자열

2. s 를 왼쪽으로 x 칸만큼 회전시켰을 때, s 가 올바른 괄호 문자열이 되게 하는 
  x 의 개수를 return 하는 함수 만들기

## 풀이
1. s-1 번만큼 왼쪽으로 rotate 했을 때, 올바른 괄호 문자열이 되는지 체크하기
"""

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    
    static List<Character> stack = new ArrayList<>();
    
    static int cnt = 0;
    
    static void checkRotate(int i) {
        ArrayDeque<Character> stack2 = new ArrayDeque<>();
    
        Collections.rotate(stack, -i);
        //System.out.println("stack : " + stack);
        
        for (char c : stack) {
            if (stack2.isEmpty()) {
                stack2.add(c);
            } else {
                char sc = stack2.pollLast();
                if (sc == '(' && c == ')') {
                    continue;
                } else if (sc == '[' && c == ']') {
                    continue;
                } else if (sc == '{' && c == '}') {
                    continue;
                } else {
                    stack2.add(sc);
                    stack2.add(c);
                }
            }
            //System.out.println("현재 스택 : " + stack2);
        }
        //System.out.println("stack2 : " + stack2);
        
        if (stack2.isEmpty()) {
            cnt++;
        }
        Collections.rotate(stack, i);
    }
    
    public int solution(String s) {
        
        char[] arr = s.toCharArray();
        
        for (char c : arr) {
            stack.add(c);
        }
        
        for (int i = 0; i < stack.size(); i++) {
            checkRotate(i);
        }
        
        return cnt;
    }
}